core/drivers/crypto/caam/hash/caam_hash.c

```diff
@@ -385,183 +470,98 @@ TEE_Result caam_hash_hmac_update(struct hashctx *ctx, const uint8_t *data,
-				caam_desc_add_word(desc, src_sgt.length);
-			} else {
-				caam_desc_add_word(desc,
-						   FIFO_LD_SGT(CLASS_2, MSG,
-							       LAST_C2,
-							       src_sgt.length));
-				caam_desc_add_ptr(desc,
-						  virt_to_phys(src_sgt.sgt));
-			}
-			caam_sgt_cache_op(TEE_CACHECLEAN, &src_sgt);
-		} else {
-			if (src_sgt.length > FIFO_LOAD_MAX) {
-				caam_desc_add_word(desc,
-						   FIFO_LD_EXT(CLASS_2, MSG,
-							       LAST_C2));
-				caam_desc_add_ptr(desc, src_sgt.buf->paddr);
-				caam_desc_add_word(desc, src_sgt.length);
-			} else {
-				caam_desc_add_word(desc,
-						   FIFO_LD(CLASS_2, MSG,
-							   LAST_C2,
-							   src_sgt.length));
-				caam_desc_add_ptr(desc, src_sgt.buf->paddr);
-			}
-
-			if (!src_sgt.buf->nocache)
-				cache_operation(TEE_CACHECLEAN,
-						src_sgt.buf->data,
-						src_sgt.length);
-		}
+		goto save_posted;
+	}
 
-		ctx->blockbuf.filled = 0;
+	ret = caam_dmaobj_init_input(&src, data, size_inmade);
+	if (ret)
+		goto exit_update;
 
-		/* Save the running context */
-		caam_desc_add_word(desc,
-				   ST_NOIMM(CLASS_2, REG_CTX, ctx->ctx.length));
-		caam_desc_add_ptr(desc, ctx->ctx.paddr);
+	ret = caam_dmaobj_prepare(&src, NULL, alg->size_block);
+	if (ret)
+		goto exit_update;
 
-		HASH_DUMPDESC(desc);
+	size_todo = size_inmade;
 
-		/* Ensure Context register data are not in cache */
-		cache_operation(TEE_CACHEINVALIDATE, ctx->ctx.data,
-				ctx->ctx.length);
+	for (offset = 0; offset < size_inmade;
+	     offset += size_done, size_todo -= size_done) {
+		size_done = size_todo;
+		HASH_TRACE("Do input %zu bytes, offset %zu", size_done, offset);
 
-		jobctx.desc = desc;
-		retstatus = caam_jr_enqueue(&jobctx, NULL);
+		ret = caam_dmaobj_sgtbuf_build(&src, &size_done, offset,
+					       alg->size_block);
+		if (ret)
+			goto exit_update;
 
-		if (retstatus == CAAM_NO_ERROR) {
-			ret = TEE_SUCCESS;
-			HASH_DUMPBUF("CTX", ctx->ctx.data, ctx->ctx.length);
-		} else {
-			HASH_TRACE("CAAM Status 0x%08" PRIx32, jobctx.status);
-			ret = job_status_to_tee_result(jobctx.status);
+		/*
+		 * Need to re-adjust the length of the data if the
+		 * posted data block is not empty and the SGT/Buffer
+		 * is part of the full input data to do.
+		 */
+		if (ctx->blockbuf.filled && size_done < size_todo) {
+			size_done -= ctx->blockbuf.filled;
+			src.sgtbuf.length = size_done;
 		}
-	} else {
-		ret = TEE_SUCCESS;
 
-		if (size_topost) {
-			/* All input data must be saved */
-			size_inmade = 0;
-		}
+		ret = do_update_hash(ctx, &src);
+		if (ret)
+			goto exit_update;
 	}
 
+save_posted:
 	if (size_topost && data) {
-		/*
-		 * Set the full data size of the input buffer.
-		 * indata.length has been changed when creating the SGT
-		 * object.
-		 */
-		indata.length = len;
 		HASH_TRACE("Posted %zu of input len %zu made %zu", size_topost,
-			   indata.length, size_inmade);
-		ret = caam_cpy_block_src(&ctx->blockbuf, &indata, size_inmade);
+			   len, size_inmade);
+
```

Comment:
Please keep declarations of variables at the beginning of a block.

Reply:
Ok